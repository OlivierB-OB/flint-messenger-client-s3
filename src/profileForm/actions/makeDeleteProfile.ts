import axios from 'axios';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { IAppState } from '../../appReducer';
import { history } from '../../history';
import { updateProfileFormStatus } from './updateProfileFormStatus';

export function makeDeleteProfile() {
  return async (dispatch: ThunkDispatch<IAppState, void, Action>, getState: () => IAppState) => {
    
    // FIXME confirm before sending

    dispatch(updateProfileFormStatus('unavailable'));

    try {
      await axios.delete(`${process.env.REACT_APP_BACKEND}/profile`, { withCredentials: true });
      history.push(`/login`);
    } catch (error) {
      dispatch(updateProfileFormStatus('error'));
    }
  };
}
