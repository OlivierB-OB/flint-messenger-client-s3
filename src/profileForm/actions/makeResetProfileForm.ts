import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { action } from '../../utils/action';
import { IAppState } from '../../appReducer';
import { updateProfileFormStatus } from './updateProfileFormStatus';
import { resetProfileForm } from './resetProfileForm';

export const makeResetProfileForm = action(() => {
  return (dispatch: ThunkDispatch<IAppState, void, Action>, getState: () => IAppState) => {
    const { info } = getState().identity;
    if (!info) dispatch(updateProfileFormStatus('unavailable'));
    else dispatch(resetProfileForm(info));
  };
});
