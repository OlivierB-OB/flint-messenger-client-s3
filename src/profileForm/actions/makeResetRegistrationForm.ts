import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { IAppState } from '../../appReducer';
import { resetProfileForm } from './resetProfileForm';

export function makeResetRegistrationForm() {
  return (dispatch: ThunkDispatch<IAppState, void, Action>) => {
    dispatch(resetProfileForm());
  };
}
