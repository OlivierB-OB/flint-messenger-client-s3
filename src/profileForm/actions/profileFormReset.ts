import { IProfileFormResetAction, PROFILE_FORM_RESET } from '../types';

export function profileFormReset(): IProfileFormResetAction {
  return { type: PROFILE_FORM_RESET };
}
