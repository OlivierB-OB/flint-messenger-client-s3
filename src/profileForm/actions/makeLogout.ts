import axios from 'axios';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { action } from '../../utils/action';
import { IAppState } from '../../appReducer';
import { updateProfileFormStatus } from './updateProfileFormStatus';
import { makeExitApplication } from '../../layout/actions/makeExitApplication';
import { config } from '../../config';

const { api_backend_url } = config;

export const makeLogout = action(() => {
  return async (dispatch: ThunkDispatch<IAppState, void, Action>) => {
    dispatch(updateProfileFormStatus('unavailable'));

    try {
      await axios.post(`${api_backend_url}/logout`, { withCredentials: true });
      dispatch(makeExitApplication());
    } catch (error) {
      dispatch(updateProfileFormStatus('error'));
    }
  };
});
