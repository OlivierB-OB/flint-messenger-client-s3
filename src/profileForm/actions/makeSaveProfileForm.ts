import axios from 'axios';
import { batch } from 'react-redux';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { action } from '../../utils/action';
import { IAppState } from '../../appReducer';
import { updateProfileFormStatus } from './updateProfileFormStatus';
import { updateIdentity } from '../../identity/actions/updateIdentity';
import { config } from '../../config';

const { api_backend_url } = config;

export const makeSaveProfileForm = action(() => {
  return async (dispatch: ThunkDispatch<IAppState, void, Action>, getState: () => IAppState) => {
    dispatch(updateProfileFormStatus('unavailable'));

    const { profileForm } = getState();
    const { lastName, firstName, password } = profileForm.fields;

    // FIXME validate form before sending

    try {
      const data = {
        lastName: lastName.value,
        firstName: firstName.value,
        password: password.value,
      };
      if (!password.value) delete data.password;
      const response = await axios.patch(`${api_backend_url}/profile`, data, { withCredentials: true });
      batch(() => {
        dispatch(updateIdentity(response.data));
        dispatch(updateProfileFormStatus('success'));
      });
    } catch (error) {
      dispatch(updateProfileFormStatus('error'));
    }
  };
});
