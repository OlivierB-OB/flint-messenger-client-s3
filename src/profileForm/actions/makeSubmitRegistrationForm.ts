import axios from 'axios';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { action } from '../../utils/action';
import { IAppState } from '../../appReducer';
import { history } from '../../history';
import { updateProfileFormStatus } from './updateProfileFormStatus';
import { makeInitializeApplication } from '../../layout/actions/makeInitializeApplication';
import { config } from '../../config';
import { makeExitApplication } from '../../layout/actions/makeExitApplication';

const { api_backend_url } = config;

export const makeSubmitRegistrationForm = action(() => {
  return async (dispatch: ThunkDispatch<IAppState, void, Action>, getState: () => IAppState) => {
    dispatch(updateProfileFormStatus('unavailable'));

    const { profileForm } = getState();
    const { email, lastName, firstName, password } = profileForm.fields;

    // FIXME validate form before sending

    try {
      const response = await axios.post(
        `${api_backend_url}/register`,
        {
          email: email.value,
          lastName: lastName.value,
          firstName: firstName.value,
          password: password.value,
        },
        { withCredentials: true },
      );
      await dispatch(makeExitApplication());
      dispatch(makeInitializeApplication(response.data));
      history.push(`/profile`);
    } catch (error) {
      dispatch(updateProfileFormStatus('error'));
    }
  };
});
