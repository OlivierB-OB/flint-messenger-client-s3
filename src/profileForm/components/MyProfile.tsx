import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import Container from '@material-ui/core/Container';
import Grid from '@material-ui/core/Grid';
import React, { useEffect } from 'react';
import { Action } from 'redux';
import { connect } from 'react-redux';
import { ThunkDispatch } from 'redux-thunk';
import { IAppState } from '../../appReducer';
import { makeResetProfileForm } from '../actions/makeResetProfileForm';
import { makeSaveProfileForm } from '../actions/makeSaveProfileForm';
import { updateProfileForm } from '../actions/updateProfileForm';
import { IProfileFormFields, IProfileFormStatus } from '../types';
import { IdentitySection } from './IdentitySection';
import { CredentialsSection } from './CredentialsSection';
import { IIdentityStatus } from '../../identity/types';
import { Loading } from '../../layout/components/Loading';
import { Alert } from '../../layout/components/Alert';
import { makeDeleteProfile } from '../actions/makeDeleteProfile';

export interface IProfileFormProps {
  identityStatus: IIdentityStatus;
  formStatus: IProfileFormStatus;
  fields: IProfileFormFields;
  update<T extends keyof IProfileFormFields>(field: T, value: IProfileFormFields[T]['value']): void;
  saveProfile(): void;
  resetProfile(): void;
  deleteProfile(): void;
}

export function ProfileForm({
  identityStatus,
  formStatus,
  fields,
  update,
  resetProfile,
  saveProfile,
  deleteProfile,
}: IProfileFormProps) {
  const { email, firstName, lastName, password, confirmation } = fields;
  useEffect(() => void resetProfile(), [identityStatus, resetProfile]);
  if ([identityStatus, formStatus].includes('unavailable')) return <Loading />;
  return (
    <Container>
      <Box style={{ margin: '2rem 0' }}>
        <Alert status={formStatus} success="Profile successfully updated!" />
      </Box>
      <Box style={{ margin: '2rem 0' }}>
        <Grid container justify="flex-end">
          <Grid item xs={2}>
            <Button variant="contained" color="secondary" fullWidth={true} onClick={deleteProfile}>
              Delete account
            </Button>
          </Grid>
        </Grid>
      </Box>
      <form onSubmit={saveProfile}>
        <Box style={{ margin: '2rem 0' }}>
          <Grid container justify="space-evenly" alignItems="flex-start">
            <Grid item xs={4}>
              <IdentitySection
                allowEmailEdition={false}
                email={email}
                firstName={firstName}
                lastName={lastName}
                update={update}
              />
            </Grid>
            <Grid item xs={4}>
              <CredentialsSection password={password} confirmation={confirmation} update={update} />
            </Grid>
          </Grid>
        </Box>
        <Box style={{ margin: '2rem 0' }}>
          <Grid container justify="space-between">
            <Grid item xs={2}>
              <Button variant="contained" fullWidth={true} onClick={resetProfile}>
                Reset
              </Button>
            </Grid>
            <Grid item xs={2}>
              <Button variant="contained" color="primary" fullWidth={true} type="submit">
                Save
              </Button>
            </Grid>
          </Grid>
        </Box>
      </form>
    </Container>
  );
}

const mapStateToProps = ({ identity, profileForm }: IAppState) => ({
  identityStatus: identity.status,
  formStatus: profileForm.status,
  fields: profileForm.fields,
});

const mapDispatchToProps = (dispatch: ThunkDispatch<IAppState, void, Action>) => ({
  update: <T extends keyof IProfileFormFields>(field: T, value: IProfileFormFields[T]['value']) =>
    dispatch(updateProfileForm(field, value)),
  resetProfile: () => dispatch(makeResetProfileForm()),
  saveProfile: () => dispatch(makeSaveProfileForm()),
  deleteProfile: () => dispatch(makeDeleteProfile()),
});

export const MyProfile = connect(mapStateToProps, mapDispatchToProps)(ProfileForm);
