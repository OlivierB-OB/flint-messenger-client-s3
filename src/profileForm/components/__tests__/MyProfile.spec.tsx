import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import { MyProfile } from '../MyProfile';
import { makeResetProfileForm } from '../../actions/makeResetProfileForm';
import { makeSaveProfileForm } from '../../actions/makeSaveProfileForm';
import { makeDeleteProfile } from '../../actions/makeDeleteProfile';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('MyProfile', () => {
  let Connected: any = MyProfile;
  let spyReset: jest.Mock;
  let restoreReset: () => void;
  let spySave: jest.Mock;
  let restoreSave: () => void;
  let spyDelete: jest.Mock;
  let restoreDelete: () => void;

  beforeEach(() => {
    [spyReset, restoreReset] = makeResetProfileForm.mock(jest.fn());
    [spySave, restoreSave] = makeSaveProfileForm.mock(jest.fn());
    [spyDelete, restoreDelete] = makeDeleteProfile.mock(jest.fn());
  });

  afterEach(() => {
    restoreReset();
    restoreSave();
    restoreDelete();
  });

  it('should provide status, fields and callbacks to ProfileForm', async () => {
    const store = mockStore({
      identity: {
        status: 'identity status',
      },
      profileForm: {
        status: 'profileForm status',
        fields: 'profileForm fields',
      },
    });
    const component = shallow(<Connected store={store} />).dive();
    expect(component).toMatchSnapshot();

    (component.prop('update') as any)('field to update', 'new value');
    expect(store.getActions()).toEqual([
      {
        field: 'field to update',
        value: 'new value',
        type: 'UPDATE_PROFILE_FORM',
      },
    ]);
    store.clearActions();

    (component.prop('resetProfile') as any)();
    expect(spyReset).toHaveBeenCalled();

    (component.prop('saveProfile') as any)();
    expect(spySave).toHaveBeenCalled();

    (component.prop('deleteProfile') as any)();
    expect(spyDelete).toHaveBeenCalled();
  });
});
