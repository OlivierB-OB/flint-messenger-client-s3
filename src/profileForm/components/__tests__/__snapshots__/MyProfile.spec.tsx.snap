// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyProfile should provide status, fields and callbacks to ProfileForm 1`] = `
<ProfileForm
  deleteProfile={[Function]}
  fields="profileForm fields"
  formStatus="profileForm status"
  identityStatus="identity status"
  resetProfile={[Function]}
  saveProfile={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  update={[Function]}
/>
`;
